tensorflow/python/framework/python_op_gen.cc

```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
   return prelude_ + result_;
 }
 
+string GenEagerPythonOp::GetTypeAnnotatedParams() {
+  // holds mappings from param name to its type annotation
+  std::unordered_map<string, string> param_type_map;
+  for (int i = 0; i<op_def_.attr_size(); ++i) {
+    const auto& attr(op_def_.attr(i));
+    if (attr.type() == "type") {
+      bool has_dtype_half = false;
+      for (int t : attr.allowed_values().list().type()) {
+        if (t == 19) { // DT_HALF = 19
+          has_dtype_half = true;
+          break;
+        }
+      }
+
+      // Do not add type annotations to an arg that can be of type half
+      if (has_dtype_half) continue;
+
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
+      param_type_map[attr.name()] = type_var_name;
+    }
+  }
+
+  for (int i = 0; i < api_def_.arg_order_size(); ++i) {
```

Comment:
Does this need to be done in arg order? Could we just enumerate over op_def_.input_arg instead?